src/modules/lost/functions.c

```diff
@@ -405,6 +522,229 @@ int lost_held_function(struct sip_msg *_m, char *_con, char *_pidf, char *_url,
 	return LOST_CLIENT_ERROR;
 }
 
+
+/*
+ * lost_held_dereference(msg, url, pidf, err, rtime, rtype)
+ * assembles and runs HELD locationRequest (POST), returns result as pidf
+ */
+int lost_held_dereference(struct sip_msg *_m, char *_url, char *_pidf,
+		char *_err, char *_rtime, char *_rtype)
+{
+	pv_spec_t *pspidf;
+	pv_spec_t *pserr;
+
+	pv_value_t pvpidf;
+	pv_value_t pverr;
+
+	xmlDocPtr doc = NULL;
+	xmlNodePtr root = NULL;
+
+	str url = STR_NULL;
+	str rtm = STR_NULL;
+	str rtp = STR_NULL;
+
+	str res = STR_NULL; /* return value location response */
+	str err = STR_NULL; /* return value error */
+
+	char *ptr = NULL;
+	char *lisurl = NULL;
+	char *heldreq = NULL;
+	char *rtype = NULL;
+
+	long ltime = 0;
+
+	int len = 0;
+	int curl = 0;
+	int exact = 0;
+
+	if(_url == NULL || _rtime == NULL || _pidf == NULL || _rtype == NULL
+			|| _err == NULL) {
+		LM_ERR("invalid parameter\n");
+		goto err;
+	}
+
+	/* dereference url from parameter */
+	if(_url) {
+		if(get_str_fparam(&url, _m, (gparam_p)_url) != 0) {
+			LM_ERR("cannot get dereference url\n");
+			goto err;
+		}
+		if(url.len == 0) {
+			LM_ERR("no dereference url found\n");
+			goto err;
+		}
+	}
+
+	/* response time from parameter */
+	if(_rtime) {
+		if(get_str_fparam(&rtm, _m, (gparam_p)_rtime) != 0) {
+			LM_ERR("cannot get response time\n");
+			goto err;
+		}
+		if(rtm.len == 0) {
+			LM_WARN("no response time found\n");
+			ltime = 0;
+		} else {
+			ltime = strtol(rtm.s, &ptr, 10);
+		}
+	}
+
+	/* response type from parameter */
+	if(_rtype) {
+		if(get_str_fparam(&rtp, _m, (gparam_p)_rtype) != 0) {
+			LM_ERR("cannot get response type\n");
+			goto err;
+		}
+		if(rtp.len == 0) {
+			LM_WARN("no response type found\n");
+			rtype = NULL;
+		} else {
+			len = 0;
+			/* response type string sanity check */
+			rtype = lost_held_type(rtp.s, &exact, &len);
+			if(len == 0) {
+				LM_WARN("cannot normalize [%.*s]\n", rtp.len, rtp.s);
+				rtype = NULL;
+			}
+		}
+	}
+
+	/* default responseTime: emergencyRouting */
+	heldreq = lost_held_post_request(&len, 0, rtype);
+
+	/* responseTime: milliseconds */
+	if((ltime > 0) && (strlen(ptr) == 0)) {
+		heldreq = lost_held_post_request(&len, ltime, rtype);
+	}
+
+	/* responseTime: emergencyRouting|emergencyDispatch */
+	if((ltime == 0) && (strlen(ptr) > 0)) {
+		if(strncasecmp(ptr, HELD_ED, strlen(HELD_ED)) == 0) {
+			heldreq = lost_held_post_request(&len, -1, rtype);
+		}
+		if(strncasecmp(ptr, HELD_ER, strlen(HELD_ER)) == 0) {
+			heldreq = lost_held_post_request(&len, 0, rtype);
+		}
+	}
+
+	/* clean up */
+	if(rtype != NULL) {
+		pkg_free(rtype);
+	}
+
+	if(heldreq != NULL && len == 0) {
+		LM_ERR("could not create POST request\n");
+		goto err;
+	}
+
+	LM_DBG("POST request: [%.*s]\n", len, heldreq);
+
+	/* curl doesn't like str */
+	len = 0;
+	lisurl = lost_copy_string(url, &len);
```

Comment:
same as above

Reply:
curl requires a zero terminated string as url, which is not added by the http_client API query function `http_client_query_c`